src/dmd/typesem.d

```diff
@@ -3223,6 +3223,10 @@ private extern(C++) final class DotExpVisitor : Visitor
                 }
                 checkAccess(e.loc, sc, null, v);
                 Expression ve = new VarExp(e.loc, v);
+                if (!isTrivialExp(e))
```

Comment:
@FeepingCreature - you could also use `hasSideEffects(e)` (or `e.hasSideEffects()` if you prefer UFCS).  When having an initial glance at the implementation, it didn't seem like it was a strong enough guarantee, but having a closer look, only strongly pure functions would have their calls omitted here.  Which is harmless.

Reply:
As far as I can tell, `pure void foo() { while (true) { } }` is strongly pure, but must under no circumstances be optimized away.

Reply:
Ah, bugger those looping pure functions. :-)

Yep, makes sense, and LGTM.

Reply:
@FeepingCreature - For brownie points, you can add a comment with a link to the bugzilla reference.